pkg/autoupdate/tuf/autoupdate.go

```diff
@@ -164,73 +161,84 @@ func (ta *TufAutoupdater) checkForUpdate() error {
-			"release_version", ta.versionFromTarget(custom.Target),
-			"binary", ta.binary,
+			"release_version", ta.versionFromTarget(custom.Target, binary),
+			"binary", binary,
 			"channel", ta.channel,
 		)
 
-		break
+		return nil
 	}
 
-	return nil
+	return fmt.Errorf("expected release file %s for binary %s to be in targets but it was not", targetReleaseFile, binary)
 }
 
-func (ta *TufAutoupdater) versionFromTarget(target string) string {
+func (ta *TufAutoupdater) versionFromTarget(target string, binary string) string {
 	// The target is in the form `launcher/linux/launcher-0.13.6.tar.gz` -- trim the prefix and the file extension to return the version
-	prefixToTrim := fmt.Sprintf("%s/%s/%s-", ta.binary, ta.operatingSystem, ta.binary)
+	prefixToTrim := fmt.Sprintf("%s/%s/%s-", binary, ta.operatingSystem, binary)
 
 	return strings.TrimSuffix(strings.TrimPrefix(target, prefixToTrim), ".tar.gz")
 }
 
-func (ta *TufAutoupdater) rollingErrorCount() int {
-	ta.lock.RLock()
-	defer ta.lock.RUnlock()
-
-	oneDayAgo := time.Now().Add(-24 * time.Hour).Unix()
-	errorCount := 0
-	for _, errorTimestamp := range ta.errorCounter {
-		if errorTimestamp < oneDayAgo {
-			continue
-		}
-		errorCount += 1
+func (ta *TufAutoupdater) storeError(autoupdateErr error) {
+	timestamp := strconv.Itoa(int(time.Now().Unix()))
+	if err := ta.store.Set([]byte(timestamp), []byte(autoupdateErr.Error())); err != nil {
+		level.Debug(ta.logger).Log("msg", "could store autoupdater error", "err", err)
 	}
-
-	return errorCount
 }
 
-func (ta *TufAutoupdater) cleanUpOldErrorCounts() {
-	ta.lock.Lock()
-	defer ta.lock.Unlock()
-
-	errorsWithinLastDay := make([]int64, 0)
+func (ta *TufAutoupdater) cleanUpOldErrors() {
+	// We want to delete all errors more than 1 week old
+	errorTtl := 7 * 24 * time.Hour
+
+	// Read through all keys in bucket to determine which ones are old enough to be deleted
+	keysToDelete := make([][]byte, 0)
+	if err := ta.store.ForEach(func(k, _ []byte) error {
+		// Key is a timestamp
+		ts, err := strconv.ParseInt(string(k), 10, 64)
+		if err != nil {
+			// Delete the corrupted key
+			keysToDelete = append(keysToDelete, k)
+			return fmt.Errorf("parsing key %s as timestamp: %w", string(k), err)
+		}
 
-	oneDayAgo := time.Now().Add(-24 * time.Hour).Unix()
-	for _, errorTimestamp := range ta.errorCounter {
-		if errorTimestamp < oneDayAgo {
-			continue
+		errorTimestamp := time.Unix(ts, 0)
+		if errorTimestamp.Add(errorTtl).Before(time.Now()) {
+			keysToDelete = append(keysToDelete, k)
 		}
-		errorsWithinLastDay = append(errorsWithinLastDay, errorTimestamp)
+
+		return nil
+	}); err != nil {
+		level.Debug(ta.logger).Log("msg", "could not iterate over bucket items to determine which are expired", "err", err)
 	}
 
-	ta.errorCounter = errorsWithinLastDay
+	// Delete all old keys
+	if err := ta.store.Delete(keysToDelete...); err != nil {
+		level.Error(ta.logger).Log("msg", "could not delete old autoupdater errors from bucket", "err", err)
```

Comment:
Maybe should be at `Debug` level, as long as failing to delete the key doesn't cause bad behavior.

Reply:
Yup good point, updated!